### ISSUE TITLE
[FEAT]Make download available when other downloads are in progress

### ISSUE THREAD

**Issue by Manochvarma** [MSFS-Mega-Pack/MSFS2020-livery-manager#209]
## Description

Found that when you have selected [for example: A320 Liveries] _**Select all**_ and click _**Download**_, you are not able to select other aircraft liveries to be downloaded. You need to wait until the previous selected liveries finish downloading first and then select next part to be downloaded.
## Solution

Make the download button available when download process going on at the background. The newly selected download liveries will be added to the download list.
### Why is this the best solution?
This will allow users to add download whatever liveries they wish at once rather than waiting one to be finished. Since this is big package, we can just add the list and leave it downloaded while we do other work or go for sleep (normally I will download bigger files at night time before going to sleep because 1. save time, 2. internet will be much faster).

## Additional context

None.

**Comment by davwheat** [comment]
The issue is with overloading our backend.

We don't want people downloading in parallel if we can help it at all.

We're already abusing Cloudflare's caching with TERABYTES of transfers every month. For example, in the last week, we've had 70 TB of data transferred...

Our current code in the manager has issues if we start having parallel downloads, so we'd need to rewrite it all.

**Comment by Manochvarma** [comment]
> The issue is with overloading our backend.
> 
> We don't want people downloading in parallel if we can help it at all.
> 
> We're already abusing Cloudflare's caching with TERABYTES of transfers every month. For example, in the last week, we've had 70 TB of data transferred...

Is this a free service by Cloudflare or it is payware?
> Our current code in the manager has issues if we start having parallel downloads, so we'd need to rewrite it all.

Hmm, might need to find solution for this. I understand rewriting the whole code will take bunch of time but if that gives good reputation and make the installer much professional then why not? Take your time. Are you the only one doing the installer work?

> You say that...
> 
> > Since this is big package, we can just add the list and leave it downloaded while we do other work or go for sleep
> 
> Surely that's a reason not to allow this? People can just simply select the liveries they want and install.

When you have selected few liveries (example few A320 livery) and you have clicked the download button, it wont allow you to click other liveries that you wish to download from other aircraft package. You have to wait until the first download to be completed before choosing other additional liveries to be downloaded. That's the main reason I created this feature request.

**Comment by davwheat** [comment]
We have a proper CDN provided to us free of charge by a company called Bunny CDN. They give us free credits so we can host this project.

We then reinforce that with a Cloudflare Pro plan, so that we don't waste all the credits in one day. We pay so that we can have more Page Rules and Analytics.

---

Rewriting the code would be a pain. The best way I can think of is a queue system. The code loops until no items are in the queue. Right now we just repeat it for every item in the queue instead.

If we do it until no items remain, it means we can add more items to the queue while the download system is running.

There's another guy who works on the manager with me too, but we're both still in full time education, so we need to prioritise.

I should have some time over Christmas to work on this. The main priority for this project is just to make sure things work. Then it's localisation, so it's accessible to everyone. Then it's feature requests.

I do read over all the issues and requests, even if I don't reply or interact with them.

---

Why don't people just select all the liveries they want first time and then press install?

Would adding a dialog stating "once you start downloading, it needs to complete before you can download more liveries" be a good stop-gap solution? You'd then be able to choose to "choose more liveries", "start downloading", as well as a "never show this again" checkbox.

**Comment by GewoonJaap** [comment]
Yeah I agree with Dave, you should select all the liveries you want and then press download. I also understand your feedback. However it is low priority for now, since we are both very busy with school

**Comment by Manochvarma** [comment]
> We have a proper CDN provided to us free of charge by a company called Bunny CDN. They give us free credits so we can host this project.
> 
> We then reinforce that with a Cloudflare Pro plan, so that we don't waste all the credits in one day. We pay so that we can have more Page Rules and Analytics.
> 

Ok I got it.

> Rewriting the code would be a pain. The best way I can think of is a queue system. The code loops until no items are in the queue. Right now we just repeat it for every item in the queue instead.
> 
> If we do it until no items remain, it means we can add more items to the queue while the download system is running.
> 
> There's another guy who works on the manager with me too, but we're both still in full time education, so we need to prioritise.
> 

I am glad both of you doing such a great work by helping this project even as a student. Proud of both of you!

> I should have some time over Christmas to work on this. The main priority for this project is just to make sure things work. Then it's localisation, so it's accessible to everyone. Then it's feature requests.
> 
> I do read over all the issues and requests, even if I don't reply or interact with them.
> 
> Why don't people just select all the liveries they want first time and then press install?
> 
> Would adding a dialog stating "once you start downloading, it needs to complete before you can download more liveries" be a good stop-gap solution? You'd then be able to choose to "choose more liveries", "start downloading", as well as a "never show this again" checkbox.

We can't aspect everyone should do so. In my case, I was first thinking downloading only the A320 liveries. After selecting all the A320 liveries, I clicked the download button. Then got interested seeing other airlines liveries from 787 and 747 which I did not aware of. So, decided to select that too but unfortunately the software not allowing me to do so. Therefore, I waited few hours to complete the download for A320 and then reselected the other liveries.
